include/deal.II/cgal/triangulation.h

```diff
@@ -0,0 +1,403 @@
+  void
+  cgal_surface_to_dealii_triangulation(
+    const CGAL_MeshType &         cgal_mesh,
+    Triangulation<dim, spacedim> &triangulation)
+  {
+    Assert(triangulation.n_cells() == 0,
+           ExcMessage(
+             "Triangulation must be empty upon calling this function."));
+    Assert(dim == 2 && spacedim == 3,
+           ExcMessage("This function works only with dim==2 and spacedim==3."));
+
+
+    auto is_surface_mesh =
+      boost::hana::is_valid([](auto &&obj) -> decltype(obj.faces()) {});
+
+    auto is_polyhedral =
+      boost::hana::is_valid([](auto &&obj) -> decltype(obj.facets_begin()) {});
+
+    // Collect Vertices
+    std::size_t                          n_cgal_vertices;
+    std::vector<dealii::Point<spacedim>> vertices;
+    std::vector<CellData<dim>>           cells;
+    SubCellData                          subcell_data;
+    if constexpr (is_surface_mesh(cgal_mesh))
+      {
+        n_cgal_vertices = cgal_mesh.num_vertices();
+        vertices.reserve(n_cgal_vertices);
+        for (const auto &p : cgal_mesh.points())
+          {
+            vertices.emplace_back(
+              CGALWrappers::cgal_point_to_dealii_point<spacedim>(p));
+          }
+      }
+    else if constexpr (is_polyhedral(cgal_mesh))
+      {
+        n_cgal_vertices = cgal_mesh.size_of_vertices();
+        vertices.reserve(n_cgal_vertices);
+        for (auto it = cgal_mesh.points_begin(); it != cgal_mesh.points_end();
+             ++it)
+          {
+            vertices.emplace_back(
+              CGALWrappers::cgal_point_to_dealii_point<spacedim>(*it));
+          }
+      }
+
+
+    // Different loops for Polyhedron or Surface_mesh types
+    if constexpr (is_surface_mesh(cgal_mesh))
+      {
+        const unsigned int vertices_per_face =
+          CGAL::vertices_around_face(
+            cgal_mesh.halfedge(*(cgal_mesh.faces().begin())), cgal_mesh)
+            .size();
+
+        // Collect CellData
+        for (const auto &face : cgal_mesh.faces())
+          {
+            CellData<dim> c(vertices_per_face);
+            auto          it_vertex = c.vertices.begin();
+            for (const auto v :
+                 CGAL::vertices_around_face(cgal_mesh.halfedge(face),
+                                            cgal_mesh))
+              {
+                *(it_vertex++) = v;
+              }
+
+            if (vertices_per_face == 4)
+              std::swap(c.vertices[3], c.vertices[2]);
+
+            cells.emplace_back(c);
+          }
+        GridTools::delete_unused_vertices(vertices, cells, subcell_data);
+        GridTools::consistently_order_cells(cells);
+        triangulation.create_triangulation(vertices, cells, subcell_data);
+      }
+    else if constexpr (is_polyhedral(cgal_mesh))
+      {
+        const unsigned int vertices_per_face =
+          cgal_mesh.facets_begin()->facet_degree();
+
+        auto cgal_mesh_nc = const_cast<CGAL_MeshType *>(
```

Comment:
I'm not sure I understand what you are trying to achieve here. We don't want to assume that the vertex contains an ID (this would work only on a subset of the mesh and polyhedron classes). Instead, we should map vertex(_handles) to dealii indices in a map (take a look at the function above https://github.com/dealii/dealii/pull/13664/files#diff-174da3832287deafdf57eb886cf0a30e787b3742b40d945331573530e65feb84R180-R191), and use the map to deduce the index. This is slightly less efficient, but much simpler, and works for all polyhedron/surface_mesh types.